doc/api/async_hooks.md

```diff
@@ -0,0 +1,671 @@
+// current execution scope to fire.
+const tid = aysnc_hooks.triggerId();
+
+// Create a new instance of AsyncHook. All of these callbacks are optional.
+const asyncHook = async_hooks.createHook({ init, before, after, destroy });
+
+// Allow callbacks of this AsyncHook instance to fire. This is not an implicit
+// action after running the constructor, and must be explicitly run to begin
+// executing callbacks.
+asyncHook.enable();
+
+// Disable listening for new asynchronous events. Though this will not prevent
+// callbacks from firing on asynchronous chains that have already run within
+// the scope of an enabled AsyncHook instance.
+asyncHook.disable();
+
+//
+// The following are the callbacks that can be passed to createHook().
+//
+
+// init() is called during object construction. The resource may not have
+// completed construction when this callback runs. So all fields of the
+// resource referenced by "id" may not have been populated.
+function init(id, type, triggerId, resource) { }
+
+// before() is called just before the resource's callback is called. It can be
+// called 0-N times for handles (e.g. TCPWrap), and should be called exactly 1
+// time for requests (e.g. FSReqWrap).
+function before(id) { }
+
+// after() is called just after the resource's callback has finished, and will
+// always fire. Except in the case of an uncaught exception.
+function after(id) { }
+
+// destroy() is called when an AsyncWrap instance is destroyed. In cases like
+// HTTPParser where the resource is reused, or timers where the handle is only
+// a JS object, destroy() will be triggered manually soon after after() has
+// completed.
+function destroy(id) { }
+```
+
+#### `async_hooks.createHook(callbacks)`
+
+Registers functions to be called for different lifetime events of each async
+operation.
+The above callbacks registered via an `AsyncHooks` instance fire during specific
+asynchronous events in the lifetime of the event loop. The focal point of these
+calls centers around the lifetime of the `AsyncWrap` C++ class. These callbacks
+will also be called to emulate the lifetime of handles and requests that do not
+fit this model. For example, `HTTPParser` instances are recycled to improve
+performance. So the `destroy()` callback would be called manually after a
+connection is done using it, just before it's placed back into the unused
+resource pool.
+
+All callbacks are optional. So, for example, if only resource cleanup needs to
+be tracked then only the `destroy()` callback needs to be passed. The
+specifics of all functions that can be passed to `callbacks` is in the section
+`Hook Callbacks`.
+
+**Error Handling**: If any `AsyncHook` callbacks throw, the application will
+print the stack trace and exit. The exit path does follow that of any uncaught
+exception, except for the fact that it is not catchable by an uncaught
+exception handler, so any `'exit'` callbacks will fire. Unless the application
```

Comment:
maybe remove `, except for the fact that it is not catchable by an uncaught exception handler`

and edit `, however `'exit'` callbacks will still fire.`